libsolidity/codegen/ir/IRGeneratorForStatements.cpp

```diff
@@ -1457,8 +1457,18 @@ void IRGeneratorForStatements::declareAssign(IRVariable const& _lhs, IRVariable
 }
 void IRGeneratorForStatements::declare(IRVariable const& _var)
 {
-	if (_var.type().sizeOnStack() > 0)
-		m_code << "let " << _var.commaSeparatedList() << "\n";
+	auto const& type = _var.type();
+	if (type.sizeOnStack() == 0)
+		return;
+
+	if (auto const* arrayType = dynamic_cast<ArrayType const*>(&type))
+		if (!arrayType->isDynamicallySized())
+		{
+			m_code << "let " << _var.part("mpos").name() << " := " << m_utils.allocateMemoryArrayFunction(*arrayType) << "(" << arrayType->length() << ")\n";
```

Comment:
My impression is that the declare function is too low-level to generate an assignment.

Reply:
The function `declareAssign` above it does the same when the rhs is available, not sure where it'd be better

Reply:
sure, but it is still low-level. The IRVariable mechanism should not be concerned with the semantics of solidity types.

Reply:
Updated by adding a higher level function called by `visit(VariableDeclarationStatement)`, which checks the Solidity type and either declares the var using the allocation function or calls `declare` accordingly.

Reply:
Are you doing the same with return parameters?

Reply:
good catch

Reply:
Updated